<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 5 The GSEA method | Gene Set Enrichment Analysis with R and Bioconductor</title>
<meta name="author" content="Zuguang Gu">
<meta name="generator" content="bookdown 0.27 with bs4_book()">
<meta property="og:title" content="Chapter 5 The GSEA method | Gene Set Enrichment Analysis with R and Bioconductor">
<meta property="og:type" content="book">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 5 The GSEA method | Gene Set Enrichment Analysis with R and Bioconductor">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.4.0/transition.js"></script><script src="libs/bs3compat-0.4.0/tabs.js"></script><script src="libs/bs3compat-0.4.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<meta name="description" content="5.1 Overview The tool GSEA is the mostly used for gene set enrichment analysis. In a study, genes are very moderate change, that after filter by p-values from DE anlaysis, no signficant genes are...">
<meta property="og:description" content="5.1 Overview The tool GSEA is the mostly used for gene set enrichment analysis. In a study, genes are very moderate change, that after filter by p-values from DE anlaysis, no signficant genes are...">
<meta name="twitter:description" content="5.1 Overview The tool GSEA is the mostly used for gene set enrichment analysis. In a study, genes are very moderate change, that after filter by p-values from DE anlaysis, no signficant genes are...">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Gene Set Enrichment Analysis with R and Bioconductor</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About</a></li>
<li><a class="" href="introduction.html"><span class="header-section-number">2</span> Introduction</a></li>
<li><a class="" href="gene-set-databases.html"><span class="header-section-number">3</span> Gene Set Databases</a></li>
<li><a class="" href="over-representation-analysis.html"><span class="header-section-number">4</span> Over-Representation Analysis</a></li>
<li><a class="active" href="the-gsea-method.html"><span class="header-section-number">5</span> The GSEA method</a></li>
<li><a class="" href="gsea-framework.html"><span class="header-section-number">6</span> GSEA framework</a></li>
<li><a class="" href="gene-set-enrichment-analysis-in-genomics.html"><span class="header-section-number">7</span> Gene Set Enrichment Analysis in Genomics</a></li>
<li><a class="" href="topology-based-pathway-enrichmetn.html"><span class="header-section-number">8</span> Topology-based pathway enrichmetn</a></li>
<li><a class="" href="extensions-of-gsea.html"><span class="header-section-number">9</span> Extensions of GSEA</a></li>
<li><a class="" href="visualization.html"><span class="header-section-number">10</span> Visualization</a></li>
<li><a class="" href="clustering-and-simplifying-gsea-results.html"><span class="header-section-number">11</span> Clustering and simplifying GSEA results</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="the-gsea-method" class="section level1" number="5">
<h1>
<span class="header-section-number">5</span> The GSEA method<a class="anchor" aria-label="anchor" href="#the-gsea-method"><i class="fas fa-link"></i></a>
</h1>
<div id="overview-2" class="section level2" number="5.1">
<h2>
<span class="header-section-number">5.1</span> Overview<a class="anchor" aria-label="anchor" href="#overview-2"><i class="fas fa-link"></i></a>
</h2>
<p>The tool GSEA is the mostly used for gene set enrichment analysis. In a study, genes are very moderate change,
that after filter by p-values from DE anlaysis, no signficant genes are left. Thus ORA anlaysis cannot be applied to it.
Mootha developed a new method which takes all genes in to consideration and test teh significance of gene sets by evaluating the
accumulated effect of genes’ diffenetial expression. It successfully xxx . In this chapter xxx</p>
</div>
<div id="the-gsea-method-version-one" class="section level2" number="5.2">
<h2>
<span class="header-section-number">5.2</span> The GSEA method, version one<a class="anchor" aria-label="anchor" href="#the-gsea-method-version-one"><i class="fas fa-link"></i></a>
</h2>
<p>ORA analysis actually applies a binary conversion on genes where genes pass the cutoff are set as 1 and others are set as 0. This
binary transformation over-simplifies the problem and a lot of information are lost, e.g. the differential expression, and
genes around the cutoff can be optionally set to 1 or 0 by object choise of cutoffs.</p>
<p>The GSEA method was first proposed in 2003. Authors worked on a dataset where all genes only had intermediate expression change,
where with normal differential gene expression methods, there is no significant gene left under normal cutoffs, thus
ORA analysis is impossible to perform. Mootha developed a new method named GSEA which convert genes to the rank of the differential
expression. Compared to ORA, GSEA v1 can distingush the differnet expression changes.</p>
<pre><code>ORA, differenital expression -&gt; 1111100000
GSEA   -&gt; 1 2 3 4 5 6 ...</code></pre>
</div>
<div id="gsea-v1-step-1" class="section level2" number="5.3">
<h2>
<span class="header-section-number">5.3</span> GSEA v1, step 1<a class="anchor" aria-label="anchor" href="#gsea-v1-step-1"><i class="fas fa-link"></i></a>
</h2>
<p>To make is simply to discuss, we assume the expression data is from a two-condition comparison, e.g. tumor vs normal or treatment vs control.
We assume there are repeated samples in both conditions. The first step of GSEA is to reduce the original matrix to gene-level scores,, i.e. for a row with m1+m2 values,
reduce it to a single value which measures the differnetial expression in the conditions. In GSEA, it proposed to use the signial to noise ratio (<span class="math inline">\((\mu_1 - \mu_2)/(\sigma_1 + \sigma_2)\)</span>
which actually does not taken account of the sample size. However, larger S2N absolute value, the more differential the gene is expressed.</p>
<p>After obtaining the gene-level score, a rank transformation is applied which is applied to the gene-level scores. all genes are sorted from the highest
to the lowest to form a sorted gene list.</p>
</div>
<div id="gsae-v1-step-2" class="section level2" number="5.4">
<h2>
<span class="header-section-number">5.4</span> GSAE v1, step 2<a class="anchor" aria-label="anchor" href="#gsae-v1-step-2"><i class="fas fa-link"></i></a>
</h2>
<p>With the sorted gene list, for a specific gene set, an enrichment score can be calculated. The genes in the ranked list are put into two groups, in the gene set
and not in the gene set. Denote there are total <span class="math inline">\(n\)</span> genes, <span class="math inline">\(n_k\)</span> is the number of genes in the gene set and <span class="math inline">\(p\)</span> is a certain position in the ranked list. For
genes in the gene set, we calcualte the fraction denoted as <span class="math inline">\(F_{1p}\)</span> at the position <span class="math inline">\(p\)</span>:</p>
<p><span class="math display">\[ F_{1p} = \frac{1}{n_k}\sum_{i=1}^p I(g_i \in G) \]</span></p>
<p>where <span class="math inline">\(I()\)</span> is a identify function, <span class="math inline">\(g_i\)</span> is the gene at position <span class="math inline">\(i\)</span> and <span class="math inline">\(G\)</span> is the set of genes in the gene set. Similarly, for genes not in the gene set, we calculate a second fraction denoted as <span class="math inline">\(F_{2p}\)</span>
at position <span class="math inline">\(p\)</span>:</p>
<p><span class="math display">\[ F_{2p} = \frac{1}{n - n_k}\sum_{i=1}^p I(g_i \notin G) \]</span></p>
<p>As can be seen, <span class="math inline">\(F_{1p}\)</span> and <span class="math inline">\(F_{2p}\)</span> are actually the cummulative probability of genes in the two sets. Then at position <span class="math inline">\(p\)</span>, the difference at the two cummulative probalibyy is
calculated, and the maximal difference is defined as the enrichmetn score (ES):</p>
<p><span class="math display">\[ ES = max_{p \in {1..n}} (F_{1p} - F{2p}) \]</span></p>
<p>According to the definition, actaully the ES is exactly teh Kolmogorov-Smirnov statistic and use for test whether two CDFs are the same. Till now, since ES is a statistic
with known sources, the p-value can be directly from the KS test. However, KS test is not a powerful test and in the orignal paper, which will also be demonstrated later,
the null hypothesis is constructed via permuation-based test.</p>
<p>According to the definition, ES is actaully directional, which is always non-negative. When ES score is large, the gene sets are more enriched on the top of the ranked gene list, which shows higher expression in group 1, if we assume positive different means high expression in group 1. To obtain the down-regulation, the denifition of ES can be changed a little bit:</p>
<p><span class="math display">\[ ES_{down} = min_{p \in {1..n}} (F_{1p} - F{2p}) \]</span></p>
<p>In this case, <span class="math inline">\(ES_{down}\)</span> is non-positive and a larger absolute value mean s…</p>
<p>To capture both the two directional change,</p>
<p><span class="math display">\[ k = argmax_p(|F_{1p} - F_{2p}|) \]</span>
<span class="math display">\[ ES_{bidirectional} = |F_{1k} - F_{2k}| or\]</span></p>
</div>
<div id="permutation-based-test" class="section level2" number="5.5">
<h2>
<span class="header-section-number">5.5</span> Permutation-based test<a class="anchor" aria-label="anchor" href="#permutation-based-test"><i class="fas fa-link"></i></a>
</h2>
<p>Once we have the definition of the ES statistic which measures differential expression for genes in a gene set, we need a null hypothesis distribution of ES to calculate p-values.
KS-test can be directly used to calcualte p-values, but here GSEA uses a more robust way to construct the null distribution of ES by sample permutation.</p>
<p>In the two-contidional data, if genes in the gene set are differneital expressed to the conditions, we can say the gene expression relates to the condition. A randomly case, is
the gene expression is independent to the conditions and with the observed value we want to reject the null hypothesis. Since the input matris is normally complex, and unlike the
the t-test which can have a anatical form with the normality assumption, to obtain a analytical form of ES is difficult. However, to destroy the relation of gene expression
and conditions, a simple but power way is to random sample the sample labels or the condition labels. If fixed the matrix, but random assign the xx of which same is group 1 and which
samle is group 2. In this rando setting, it actally means the gene expression has no relation to the condition because the conditions are randomly assigned. The permuation is
a random assignment process which randomly shuffle the xxx of labels, but keeps the number of samples in the two groups unchanged.</p>
<p>sample permutation can be simply done with the <code><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html">sample()</a></code> function. Assume we have a condition desigh with 10 samples of 5 group A and 5 group B:</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cmp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html">sample</a></span><span class="op">(</span><span class="va">cmp</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">cmp</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#  [1] "A" "A" "B" "B" "B" "A" "B" "A" "A" "B"</span></span></code></pre></div>
<p>Each sample permutation generates a new experimental design where the condition is independent to gene expression, thus a ES can be calculated with step 1 and 2. teh sample permutation
will be execulted to a large number of times to gain enough “random” ES for form the null distribution. The number of permulations normally is large and it is recommended to not less than 1000. The null distribution of ES is from the random ES scores and p-value is defined as the probability of ES being euqla to or larger than the observed one:</p>
<p>The p-value can be easily calculated as:</p>
<p><span class="math display">\[ p = \frac{1}{K}\sum I(ES_r \geq ES) \]</span></p>
<p>Sample permutation is a powerful and a general way in statistics. In Chapter xx, we will demonstrate it is used to construct a non-parametric null distribution for any kind of gene set level statistics. Also in Chapter x, we will also introduce permuation data by genes.</p>
</div>
<div id="the-gsea-method-version-2" class="section level2" number="5.6">
<h2>
<span class="header-section-number">5.6</span> The GSEA method, version 2<a class="anchor" aria-label="anchor" href="#the-gsea-method-version-2"><i class="fas fa-link"></i></a>
</h2>
<p>The first version of GSEA successfully discovered affected biological functions that show moderate expression chagnes. However, in a study, reseachers pointed
several limitations of the algorithm. The main argument is under the GSEA algorithm, although it can distinguish the different levels of differential expression,
genes have equal distance to neighbouring genes in the ranked list, and there are the following two consiquences:</p>
<ol style="list-style-type: decimal">
<li>genes with higher differnetial expression are more important</li>
<li>genes in the middle of hte gene list are normally have very small change or even no change. However, an assumulation in the middle can also cause an enrichment.</li>
</ol>
<p>Due to the gene-level score is not an uniform distribution while xxx, In 2005, Subrammian imporved the GSEA algorithm by taking the gene-level score as the weight,
in this case, the differential genes are enhanced and genes with weak differneital expression are suppressed. And it forms the current standard GSEA algorithm.</p>
<p>Recall how we define <span class="math inline">\(F_{1p}\)</span> and <span class="math inline">\(F_{2p}\)</span>. The new GSEA changes the definition of <span class="math inline">\(F_{1p}\)</span> as:</p>
<p><span class="math display">\[ F_{1p} = \frac{1}{n_r}\sum_{i=1}^p I(g_i \in G) \cdot |r_j|^a \]</span>
<span class="math display">\[ where n_r = \sum_{i=1}^n I(g_i \in G) \cdot |r_j|^a \]</span></p>
<p>Note <span class="math inline">\(F_{2p}\)</span> for the genes not in the genes is unchanged. The power <span class="math inline">\(a\)</span> reflects the importance of the differential expression. By default, GSEA set <span class="math inline">\(a = 1\)</span>. When
<span class="math inline">\(a = 0\)</span>, it reduces to the original version of GSEA.</p>
<p>Similarlly the ES score is the same as in the origial version, for the one-directional or two-directinoal.</p>
<p>Now the form of ES is changed, thus, it becomes the so-called “weighted KS statistics.” It is difficult to perform the test. Again, we perform the sample-permutation
that in every permutations, the new ES score is calclated .</p>
</div>
<div id="compare-the-two-gsea" class="section level2" number="5.7">
<h2>
<span class="header-section-number">5.7</span> Compare the two GSEA<a class="anchor" aria-label="anchor" href="#compare-the-two-gsea"><i class="fas fa-link"></i></a>
</h2>
<p>It would be interesting to compare the two versions of GSEA to see how the weight helps to xxx and also the power of the test.</p>
</div>
<div id="permutations" class="section level2" number="5.8">
<h2>
<span class="header-section-number">5.8</span> Permutations<a class="anchor" aria-label="anchor" href="#permutations"><i class="fas fa-link"></i></a>
</h2>
<p>To construct the null distribution, we introduced to randonly suffle samples to break the relation of xxx. There is actually a second way to construct “random dataset”
by gene permutation, which randomly shuffle the genes or identially, randoly shuffle whether genes is in the gene set or not. This is a way to break the relation of differnetial expression
and the xx of genes in the gene sets. If assume most of the genes are not differneitally expressed, the the randomized gene sets will mostly not show differneital expression.
In sample permutation, the gene-level scores need to be calculated in every permutatin, as a comparison, in gene permutation, the gene-level scores can only
be calculated once and can be repeated used. In some of current tools which implements GSEA algorithm, teh null distribution is generated by gene permutation. Without
going to deep, we wold like to mention, the two permutation methods, although they all produce ramdom dataset, they correspond to two complete different things. and with
some specific datasets, the two xxx may give complete different p-values (or conclusions)</p>
</div>
<div id="other-aspects-of-gsea" class="section level2" number="5.9">
<h2>
<span class="header-section-number">5.9</span> Other aspects of GSEA<a class="anchor" aria-label="anchor" href="#other-aspects-of-gsea"><i class="fas fa-link"></i></a>
</h2>
<div id="the-direction-of-gsea" class="section level3" number="5.9.1">
<h3>
<span class="header-section-number">5.9.1</span> The direction of GSEA<a class="anchor" aria-label="anchor" href="#the-direction-of-gsea"><i class="fas fa-link"></i></a>
</h3>
</div>
<div id="leading-edge-genes" class="section level3" number="5.9.2">
<h3>
<span class="header-section-number">5.9.2</span> Leading edge genes<a class="anchor" aria-label="anchor" href="#leading-edge-genes"><i class="fas fa-link"></i></a>
</h3>
</div>
<div id="normalized-enrichment-score" class="section level3" number="5.9.3">
<h3>
<span class="header-section-number">5.9.3</span> Normalized enrichment score<a class="anchor" aria-label="anchor" href="#normalized-enrichment-score"><i class="fas fa-link"></i></a>
</h3>
<p>Because teh enrichment score is averaged from all genes in a gene set, in general, large significant gene sets may only have a small ES scores. Thus, the ES cannot reflect
the degree of the enrichment and cannot be compariable between gene sets. The normalized</p>
</div>
</div>
<div id="compare-ora-and-gsea" class="section level2" number="5.10">
<h2>
<span class="header-section-number">5.10</span> Compare ORA and GSEA<a class="anchor" aria-label="anchor" href="#compare-ora-and-gsea"><i class="fas fa-link"></i></a>
</h2>
<p>Since ORA and GSEA are the mostly used methods for gene set enrichmetn anlaysis, it would be interesting to apply the two methods on a same dataset to compare xxx</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span></span>
<span><span class="va">gs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">F</span>, <span class="cn">T</span>, <span class="cn">T</span>, <span class="cn">T</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">T</span>, <span class="cn">T</span>, <span class="cn">T</span>, <span class="cn">T</span>, <span class="cn">F</span>, <span class="cn">T</span>, <span class="cn">T</span>, <span class="cn">T</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">T</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">T</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">gsea_cdf</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">gs</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">f1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="va">gs</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">gs</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">f2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="op">!</span><span class="va">gs</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="op">!</span><span class="va">gs</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">n</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span>, <span class="va">f1</span>, type <span class="op">=</span> <span class="st">"s"</span>, col <span class="op">=</span> <span class="fl">2</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1.1</span><span class="op">)</span>,</span>
<span>        xlab <span class="op">=</span> <span class="st">"Ranked gene list"</span>, ylab <span class="op">=</span> <span class="st">"Cumulative probability P_n(X &gt;= x)"</span>, <span class="va">...</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span>, <span class="va">f2</span>, type <span class="op">=</span> <span class="st">"s"</span>, col <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/segments.html">segments</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span>, <span class="fl">1.05</span>, <span class="fl">1</span><span class="op">:</span><span class="va">n</span>, <span class="fl">1.1</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">gs</span>, <span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">gsea_cdf</span><span class="op">(</span><span class="va">x</span>, <span class="va">gs</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="04-gsea_files/figure-html/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;"></div>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">gs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="cn">T</span>, <span class="fl">10</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="cn">F</span>, <span class="fl">30</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">gsea_cdf</span><span class="op">(</span><span class="va">x</span>, <span class="va">gs</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="04-gsea_files/figure-html/unnamed-chunk-2-2.png" width="672" style="display: block; margin: auto;"></div>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">gs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="cn">F</span>, <span class="fl">30</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="cn">T</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">gsea_cdf</span><span class="op">(</span><span class="va">x</span>, <span class="va">gs</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="04-gsea_files/figure-html/unnamed-chunk-2-3.png" width="672" style="display: block; margin: auto;"></div>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">gs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">T</span>, <span class="cn">F</span><span class="op">)</span>, <span class="fl">40</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">gsea_cdf</span><span class="op">(</span><span class="va">x</span>, <span class="va">gs</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="04-gsea_files/figure-html/unnamed-chunk-2-4.png" width="672" style="display: block; margin: auto;"></div>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">54</span><span class="op">)</span></span>
<span><span class="va">mean_diff1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="va">m1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>, mean <span class="op">=</span> <span class="va">mean_diff1</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">10</span>, mean <span class="op">=</span> <span class="op">-</span><span class="va">mean_diff1</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">m1</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"g"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="va">x1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">m1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">x2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">x1</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fa</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Group1"</span>, <span class="fl">10</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Group2"</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="va">ht_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">m1</span>, top_annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">fa</span><span class="op">)</span>, cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>, cluster_columns <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">x1</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"orange"</span>, <span class="st">"white"</span>, <span class="st">"purple"</span><span class="op">)</span>, left_annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/rowAnnotation.html">rowAnnotation</a></span><span class="op">(</span>pt1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/anno_points.html">anno_points</a></span><span class="op">(</span><span class="va">x1</span>, width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="va">x2</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"orange"</span>, <span class="st">"white"</span>, <span class="st">"purple"</span><span class="op">)</span>, left_annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/rowAnnotation.html">rowAnnotation</a></span><span class="op">(</span>pt2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/anno_points.html">anno_points</a></span><span class="op">(</span><span class="va">x2</span>, width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="va">ht_list</span>, ht_gap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"cm"</span><span class="op">)</span>, auto_adjust <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="04-gsea_files/figure-html/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;"></div>
<p>Prepare the data. Here we read three types of data:</p>
<ol style="list-style-type: decimal">
<li>the phenotype/condition data</li>
<li>the gene expression matrix</li>
<li>the gene set</li>
</ol>
<p>The gene expression data is saved in <code>.gct</code> format and experimental condition/phenotype
is saves in <code>.cls</code> format. Both formats are very simple, and you can try to write your own
code to parse them.</p>
<p>Gene sets are saved in "<code>.gmt</code> format. It is also a simple format. You can try to parse
it by your own.</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/CePa">CePa</a></span><span class="op">)</span></span>
<span><span class="va">condition</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/CePa/man/read.cls.html">read.cls</a></span><span class="op">(</span><span class="st">"data/P53.cls"</span>, treatment <span class="op">=</span> <span class="st">"MUT"</span>, control <span class="op">=</span> <span class="st">"WT"</span><span class="op">)</span><span class="op">$</span><span class="va">label</span></span>
<span><span class="va">expr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/CePa/man/read.gct.html">read.gct</a></span><span class="op">(</span><span class="st">"data/P53_collapsed_symbols.gct"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ln</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html">readLines</a></span><span class="op">(</span><span class="st">"data/c2.symbols.gmt"</span><span class="op">)</span>, <span class="st">"\t"</span><span class="op">)</span></span>
<span><span class="va">gs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">ln</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="op">-</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">gs</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">ln</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">geneset</span> <span class="op">=</span> <span class="va">gs</span><span class="op">[[</span><span class="st">"p53hypoxiaPathway"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<ul>
<li><code>expr</code></li>
<li><code>condition</code></li>
<li><code>geneset</code></li>
</ul>
<p>This gene set is very small:</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">geneset</span><span class="op">)</span></span>
<span><span class="co"># [1] 20</span></span></code></pre></div>
<p>Note here gene IDs in the expression matrix and in the gene sets are all gene
symbols, thus no more adjustment needs to be done here.</p>
<p>The gene-level difference score is set as signal-to-noise ratios, which is:</p>
<ul>
<li>mean in group 1</li>
<li>mean in group 2</li>
<li>sd in group 1</li>
<li>sd in group 2</li>
</ul>
<p><span class="math display">\[ \frac{\mu_1 - \mu_2}{\sigma_1 + \sigma_2} \]</span></p>
<p>We calculate the gene-level difference score in <code>s</code>:</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">s</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">expr</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">x1</span> <span class="op">=</span> <span class="va">x</span><span class="op">[</span><span class="va">condition</span> <span class="op">==</span> <span class="st">"WT"</span><span class="op">]</span></span>
<span>    <span class="va">x2</span> <span class="op">=</span> <span class="va">x</span><span class="op">[</span><span class="va">condition</span> <span class="op">==</span> <span class="st">"MUT"</span><span class="op">]</span></span>
<span>    <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">x1</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">x2</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span><span class="op">(</span><span class="va">x1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span><span class="op">(</span><span class="va">x2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>s: gene-level difference score</li>
</ul>
<p>Sort the gene scores from the highest to the lowest (to make it into a ranked list):</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">s</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">s</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Next we first implement the original GSEA method, which was proposed in Mootha et al., 2003.</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## original GSEA</span></span>
<span><span class="va">l_set</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">geneset</span></span>
<span><span class="va">f1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="va">l_set</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">l_set</span><span class="op">)</span></span>
<span></span>
<span><span class="va">l_other</span> <span class="op">=</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">geneset</span></span>
<span><span class="va">f2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="va">l_other</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">l_other</span><span class="op">)</span></span></code></pre></div>
<p>Here <code>f1</code> is the cumulative probability function of genes in the set and <code>f2</code> is the cumulative probability function of genes not in the set.</p>
<p>We first plot the CDF of two distributions.</p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span>, <span class="va">f1</span>, type <span class="op">=</span> <span class="st">"l"</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span>, <span class="va">f2</span>, col <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="04-gsea_files/figure-html/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;"></div>
<p>The reason why the blue locates almost on the diagonal is the gene set is very small.</p>
<p>Next the difference of cumulative probability (<code>f1 - f2</code>) at each position of the ranked gene list.
Let’s call it “the GSEA plot.”</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">f1</span> <span class="op">-</span> <span class="va">f2</span>, type <span class="op">=</span> <span class="st">"l"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fl">0</span>, lty <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">l_set</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">l_set</span><span class="op">)</span><span class="op">)</span>, pch <span class="op">=</span> <span class="st">"|"</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="04-gsea_files/figure-html/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;"></div>
<p>The enrichment score (ES) defined as <code>max(f1 - f2)</code> is:</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">es</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">f1</span> <span class="op">-</span> <span class="va">f2</span><span class="op">)</span></span>
<span><span class="va">es</span></span>
<span><span class="co"># [1] 0.2294643</span></span></code></pre></div>
<p>And the position in the “GSEA plot”:</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">f1</span> <span class="op">-</span> <span class="va">f2</span>, type <span class="op">=</span> <span class="st">"l"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fl">0</span>, lty <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">l_set</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">l_set</span><span class="op">)</span><span class="op">)</span>, pch <span class="op">=</span> <span class="st">"|"</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.max</a></span><span class="op">(</span><span class="va">f1</span> <span class="op">-</span> <span class="va">f2</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="04-gsea_files/figure-html/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;"></div>
<p>The statistic <code>es</code> actually is the Kolmogorov-Smirnov statistics, thus, we can
directly apply the KS test:</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/ks.test.html">ks.test</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">l_set</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">l_other</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># </span></span>
<span><span class="co">#   Asymptotic two-sample Kolmogorov-Smirnov test</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># data:  which(l_set) and which(l_other)</span></span>
<span><span class="co"># D = 0.22946, p-value = 0.244</span></span>
<span><span class="co"># alternative hypothesis: two-sided</span></span></code></pre></div>
<p>However, we can see the p-value is not significant, this is because KS test is not
a powerful test. Next we construct the null distribution by sample permutation.</p>
<p>In the next code chunk, the calculation of ES score is wrapped into a function,
also we use <code><a href="https://rdrr.io/r/base/colSums.html">rowMeans()</a></code> and <code><a href="https://rdrr.io/pkg/matrixStats/man/rowSds.html">rowSds()</a></code> to speed up the calculation of gene-level
scores.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/matrixStats">matrixStats</a></span><span class="op">)</span></span>
<span><span class="co"># expr: the complete expression matrix</span></span>
<span><span class="co"># condition: the condition labels of samples</span></span>
<span><span class="co"># cmp: a vector of two, cmp[1] - cmp[2] &gt; 0 means up-regulation</span></span>
<span><span class="co"># geneset: A vector of genes</span></span>
<span><span class="va">calculate_es</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">expr</span>, <span class="va">condition</span>, <span class="va">cmp</span>, <span class="va">geneset</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="va">m1</span> <span class="op">=</span> <span class="va">expr</span><span class="op">[</span>, <span class="va">condition</span> <span class="op">==</span> <span class="va">cmp</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>  <span class="co"># only samples in group 1</span></span>
<span>    <span class="va">m2</span> <span class="op">=</span> <span class="va">expr</span><span class="op">[</span>, <span class="va">condition</span> <span class="op">==</span> <span class="va">cmp</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>  <span class="co"># only samples in group 2</span></span>
<span></span>
<span>    <span class="va">s</span> <span class="op">=</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">m1</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">m2</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowSds.html">rowSds</a></span><span class="op">(</span><span class="va">m1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowSds.html">rowSds</a></span><span class="op">(</span><span class="va">m2</span><span class="op">)</span><span class="op">)</span> <span class="co"># a gene-level difference socre (S2N ratio) </span></span>
<span></span>
<span>    <span class="va">s</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">s</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>  <span class="co"># ranked gene list</span></span>
<span></span>
<span>    <span class="va">l_set</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">geneset</span></span>
<span>    <span class="va">f1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="va">l_set</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">l_set</span><span class="op">)</span>   <span class="co"># CDF for genes in the set</span></span>
<span></span>
<span>    <span class="va">l_other</span> <span class="op">=</span> <span class="op">!</span><span class="va">l_set</span></span>
<span>    <span class="va">f2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="va">l_other</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">l_other</span><span class="op">)</span>  <span class="co"># CDF for genes not in the set</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">f1</span> <span class="op">-</span> <span class="va">f2</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The ES score calculated by <code>calculate_es()</code>:</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">es</span> <span class="op">=</span> <span class="fu">calculate_es</span><span class="op">(</span><span class="va">expr</span>, <span class="va">condition</span>, cmp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"WT"</span>, <span class="st">"MUT"</span><span class="op">)</span>, geneset <span class="op">=</span> <span class="va">geneset</span><span class="op">)</span></span>
<span><span class="va">es</span></span>
<span><span class="co"># [1] 0.2294643</span></span></code></pre></div>
<p>We randomly permute sample labels or we randomly permute <code>condition</code>. We do it
1000 times. The ES scores in null distributions are saved in <code>es_rand</code>.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">es_rand</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">numeric</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">es_rand</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">=</span> <span class="fu">calculate_es</span><span class="op">(</span><span class="va">expr</span>, <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html">sample</a></span><span class="op">(</span><span class="va">condition</span><span class="op">)</span>, </span>
<span>        cmp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"WT"</span>, <span class="st">"MUT"</span><span class="op">)</span>, geneset <span class="op">=</span> <span class="va">geneset</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>p-value is calculated as the proportion of <code>es</code> being equal to or larger than in values in <code>es_rand</code>.</p>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">es_rand</span> <span class="op">&gt;=</span> <span class="va">es</span><span class="op">)</span><span class="op">/</span><span class="fl">1000</span></span>
<span><span class="co"># [1] 0.129</span></span></code></pre></div>
<p>0.129</p>
<p>The null distribution of ES:</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">es_rand</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="va">es</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="04-gsea_files/figure-html/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto;"></div>
<p>Next we implement the improved GSEA (Subramanian et al., PNAS, 2005) where gene-level scores are taken as the weight.</p>
<p>We directly modify <code>calculate_es()</code> to <code>calculate_es_v2()</code> where there is only two lines new, which we highlight in the code chunk:</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">calculate_es_v2</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">expr</span>, <span class="va">condition</span>, <span class="va">cmp</span>, <span class="va">geneset</span>, <span class="va">plot</span> <span class="op">=</span> <span class="cn">FALSE</span>, <span class="va">power</span> <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="va">m1</span> <span class="op">=</span> <span class="va">expr</span><span class="op">[</span>, <span class="va">condition</span> <span class="op">==</span> <span class="va">cmp</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="va">m2</span> <span class="op">=</span> <span class="va">expr</span><span class="op">[</span>, <span class="va">condition</span> <span class="op">==</span> <span class="va">cmp</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span>    <span class="va">s</span> <span class="op">=</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">m1</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">m2</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowSds.html">rowSds</a></span><span class="op">(</span><span class="va">m1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowSds.html">rowSds</a></span><span class="op">(</span><span class="va">m2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">s</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">s</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">l_set</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">geneset</span></span>
<span>    <span class="co"># f1 = cumsum(l_set)/sum(l_set)  # &lt;&lt;-- the original line</span></span>
<span>    <span class="va">s_set</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span><span class="op">^</span><span class="va">power</span>   <span class="co"># &lt;&lt;-- here</span></span>
<span>    <span class="va">s_set</span><span class="op">[</span><span class="op">!</span><span class="va">l_set</span><span class="op">]</span> <span class="op">=</span> <span class="fl">0</span></span>
<span>    <span class="va">f1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="va">s_set</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">s_set</span><span class="op">)</span>  <span class="co">## &lt;&lt;- here</span></span>
<span></span>
<span>    <span class="va">l_other</span> <span class="op">=</span> <span class="op">!</span><span class="va">l_set</span></span>
<span>    <span class="va">f2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="va">l_other</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">l_other</span><span class="op">)</span></span>
<span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="va">plot</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">f1</span> <span class="op">-</span> <span class="va">f2</span>, type <span class="op">=</span> <span class="st">"l"</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fl">0</span>, lty <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">l_set</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">l_set</span><span class="op">)</span><span class="op">)</span>, pch <span class="op">=</span> <span class="st">"|"</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.max</a></span><span class="op">(</span><span class="va">f1</span> <span class="op">-</span> <span class="va">f2</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">f1</span> <span class="op">-</span> <span class="va">f2</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Now we calculate the new ES score and make the GSEA plot:</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">es</span> <span class="op">=</span> <span class="fu">calculate_es_v2</span><span class="op">(</span><span class="va">expr</span>, <span class="va">condition</span>, cmp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"WT"</span>, <span class="st">"MUT"</span><span class="op">)</span>, plot <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    geneset <span class="op">=</span> <span class="va">geneset</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="04-gsea_files/figure-html/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;"></div>
<p>We can also check when <code>power = 0</code> and <code>power = 2</code>:</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">calculate_es_v2</span><span class="op">(</span><span class="va">expr</span>, <span class="va">condition</span>, cmp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"WT"</span>, <span class="st">"MUT"</span><span class="op">)</span>, plot <span class="op">=</span> <span class="cn">TRUE</span>, power <span class="op">=</span> <span class="fl">0</span>, </span>
<span>    geneset <span class="op">=</span> <span class="va">geneset</span><span class="op">)</span>  <span class="co"># same as the original GSEA</span></span>
<span><span class="co"># [1] 0.2294643</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span><span class="st">"power = 0"</span><span class="op">)</span></span>
<span><span class="fu">calculate_es_v2</span><span class="op">(</span><span class="va">expr</span>, <span class="va">condition</span>, cmp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"WT"</span>, <span class="st">"MUT"</span><span class="op">)</span>, plot <span class="op">=</span> <span class="cn">TRUE</span>, power <span class="op">=</span> <span class="fl">2</span>, </span>
<span>    geneset <span class="op">=</span> <span class="va">geneset</span><span class="op">)</span></span>
<span><span class="co"># [1] 0.8925371</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span><span class="st">"power = 2"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="04-gsea_files/figure-html/unnamed-chunk-21-1.png" width="960" style="display: block; margin: auto;"></div>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Similarly, we randomly permute samples to obtain the null distribution of ES:</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">es_rand</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">numeric</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">es_rand</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">=</span> <span class="fu">calculate_es_v2</span><span class="op">(</span><span class="va">expr</span>, <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html">sample</a></span><span class="op">(</span><span class="va">condition</span><span class="op">)</span>, </span>
<span>        cmp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"WT"</span>, <span class="st">"MUT"</span><span class="op">)</span>, geneset <span class="op">=</span> <span class="va">geneset</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The new p-value:</p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">es_rand</span> <span class="op">&gt;=</span> <span class="va">es</span><span class="op">)</span><span class="op">/</span><span class="fl">1000</span></span>
<span><span class="co"># [1] 0</span></span></code></pre></div>
<p>1/1000, &lt; 0.001</p>
<p>And the null distribution of ES:</p>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">es_rand</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="va">es</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="04-gsea_files/figure-html/unnamed-chunk-24-1.png" width="672" style="display: block; margin: auto;"></div>
<p>We can see the improved GSEA is more powerful than the original GSEA, because the
original GSEA equally weights genes and the improved GSEA weights genes based on their differential expression,
which increases the effect of diff genes. Let’s plot the weight of genes:</p>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="04-gsea_files/figure-html/unnamed-chunk-25-1.png" width="672" style="display: block; margin: auto;"></div>
<hr>
<p>Null distribution can also be constructed by gene permutation. It is very
easy to implement:</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># s: a vector of pre-calcualted gene-level scores</span></span>
<span><span class="co"># s should be sorted</span></span>
<span><span class="va">calculate_es_v2_gene_perm</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">s</span>, <span class="va">perm</span> <span class="op">=</span> <span class="cn">FALSE</span>, <span class="va">power</span> <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span></span>
<span>    </span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="va">perm</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="co"># s is still sorted, but the gene labels are randomly shuffled</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span><span class="op">)</span>  <span class="co">## &lt;&lt;- here</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="va">l_set</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">geneset</span></span>
<span>    <span class="va">s_set</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span><span class="op">^</span><span class="va">power</span></span>
<span>    <span class="va">s_set</span><span class="op">[</span><span class="op">!</span><span class="va">l_set</span><span class="op">]</span> <span class="op">=</span> <span class="fl">0</span></span>
<span>    <span class="va">f1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="va">s_set</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">s_set</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">l_other</span> <span class="op">=</span> <span class="op">!</span><span class="va">l_set</span></span>
<span>    <span class="va">f2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="va">l_other</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">l_other</span><span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">f1</span> <span class="op">-</span> <span class="va">f2</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Good thing of gene permutation is the gene-level scores only need to be calculated
once and can be repeatedly used.</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># pre-calculate gene-level scores</span></span>
<span><span class="va">m1</span> <span class="op">=</span> <span class="va">expr</span><span class="op">[</span>, <span class="va">condition</span> <span class="op">==</span> <span class="st">"WT"</span><span class="op">]</span></span>
<span><span class="va">m2</span> <span class="op">=</span> <span class="va">expr</span><span class="op">[</span>, <span class="va">condition</span> <span class="op">==</span> <span class="st">"MUT"</span><span class="op">]</span></span>
<span></span>
<span><span class="va">s</span> <span class="op">=</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">m1</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">m2</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowSds.html">rowSds</a></span><span class="op">(</span><span class="va">m1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowSds.html">rowSds</a></span><span class="op">(</span><span class="va">m2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">s</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">s</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>  <span class="co"># must be pre-sorted</span></span></code></pre></div>
<p>We calculate the null distribution of ES from gene permutation:</p>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">es</span> <span class="op">=</span> <span class="fu">calculate_es_v2_gene_perm</span><span class="op">(</span><span class="va">s</span><span class="op">)</span></span>
<span><span class="va">es_rand</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">numeric</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">es_rand</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">=</span> <span class="fu">calculate_es_v2_gene_perm</span><span class="op">(</span><span class="va">s</span>, perm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">es_rand</span> <span class="op">&gt;=</span> <span class="va">es</span><span class="op">)</span><span class="op">/</span><span class="fl">1000</span></span>
<span><span class="co"># [1] 0</span></span></code></pre></div>
<p>also &lt; 0.001</p>
<p>The null distribution of ES from gene permutation:</p>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">es_rand</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="va">es</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="04-gsea_files/figure-html/unnamed-chunk-29-1.png" width="672" style="display: block; margin: auto;"></div>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="over-representation-analysis.html"><span class="header-section-number">4</span> Over-Representation Analysis</a></div>
<div class="next"><a href="gsea-framework.html"><span class="header-section-number">6</span> GSEA framework</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#the-gsea-method"><span class="header-section-number">5</span> The GSEA method</a></li>
<li><a class="nav-link" href="#overview-2"><span class="header-section-number">5.1</span> Overview</a></li>
<li><a class="nav-link" href="#the-gsea-method-version-one"><span class="header-section-number">5.2</span> The GSEA method, version one</a></li>
<li><a class="nav-link" href="#gsea-v1-step-1"><span class="header-section-number">5.3</span> GSEA v1, step 1</a></li>
<li><a class="nav-link" href="#gsae-v1-step-2"><span class="header-section-number">5.4</span> GSAE v1, step 2</a></li>
<li><a class="nav-link" href="#permutation-based-test"><span class="header-section-number">5.5</span> Permutation-based test</a></li>
<li><a class="nav-link" href="#the-gsea-method-version-2"><span class="header-section-number">5.6</span> The GSEA method, version 2</a></li>
<li><a class="nav-link" href="#compare-the-two-gsea"><span class="header-section-number">5.7</span> Compare the two GSEA</a></li>
<li><a class="nav-link" href="#permutations"><span class="header-section-number">5.8</span> Permutations</a></li>
<li>
<a class="nav-link" href="#other-aspects-of-gsea"><span class="header-section-number">5.9</span> Other aspects of GSEA</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#the-direction-of-gsea"><span class="header-section-number">5.9.1</span> The direction of GSEA</a></li>
<li><a class="nav-link" href="#leading-edge-genes"><span class="header-section-number">5.9.2</span> Leading edge genes</a></li>
<li><a class="nav-link" href="#normalized-enrichment-score"><span class="header-section-number">5.9.3</span> Normalized enrichment score</a></li>
</ul>
</li>
<li><a class="nav-link" href="#compare-ora-and-gsea"><span class="header-section-number">5.10</span> Compare ORA and GSEA</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Gene Set Enrichment Analysis with R and Bioconductor</strong>" was written by Zuguang Gu. It was last built on 2022-09-01.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
